draft-ietf-mls-protocol.md

```diff
@@ -1000,6 +1000,70 @@ struct {
 } UserInitKey;
 ~~~~~
 
+# Message Framing
+
+[[ TODO prose ]]
+
+~~~~~
+enum {
+    invalid(0),
+    handshake(1),
+    application(2),
+    (255)
+} ContentType;
+
+struct {
+    opaque content[MLSPlaintext.length];
+    uint8 signature[MLSInnerPlaintext.sig_len];
+    uint16 sig_len;
+    ContentType type;
+    uint8 zero\_padding[length\_of\_padding];
+} MLSInnerPlaintext;
+
+struct {
+    opaque group_id<0..255>;
+    uint32 epoch;
+    uint32 sender;
+    ContentType type;
+
+    select (MLSPlaintext.type) {
+        case handshake:   Handshake;
+        case application: Application;
+    }
+
+    opaque signature<1..2^16-1>;
+} MLSPlaintext;
+
+struct {
+    opaque group_id<0..255>;
+    uint32 epoch;
+    uint32 sender;
+    uint32 generation;
+    opaque cipertext<0..2^32-1>;
+} MLSCiphertext;
```

Comment:
What do we need the sender field for in MLSCiphertext? We might even get some sort of identity hiding if we leave it out :-)

Reply:
No, indeed we don't want it outside, that is also why the signature is inside the Plaintext.

Reply:
I was misunderstanding Richard's idea at first. I believe the idea is to encrypt the `Plaintext` with different keys depending if it is `Handshake` or `Application`. In that scenario, you need the `sender` and the `generation` in the clear to know which key to pick in the case of an Application message. Note that @raphaelrobert and I are thinking on suggesting to encrypt these under the the group key to hide them if necessary, we'll discuss this at some point.

Reply:
Why do you need the sender, though? Both handshake and application keys derived from the groupkey should be sender-independent.

Reply:
Which key do you pick when receiving an application message ? :)

Reply:
I'm confused. The key for handshake encryption is derived from the `group_key`, right? So why do we need to know the sender? We only need the `group_id` and the `epoch` to derive the key if I understand correctly.